src/index.js

```diff
@@ -0,0 +1,153 @@
+  }
+}
+
+export { pick }
+
+export function useLocation() {
+  const context = React.useContext(locationContext)
+  if (!context) throw new Error('Please use a `LocationProvider` to supply a location')
+  return context.location
+}
+
+export function useNavigate() {
+  const context = React.useContext(baseContext)
+  if (!context) throw new Error('Please use a `LocationProvider` to supply a navigate function')
+  return context.navigate
+}
+
+function LocationProvider({ initialLocation = undefined, children: originalChildren }) {
+  const children = <RootBaseContextProvider children={originalChildren} />
+
+  return inBrowser
+    ? <BrowserLocationProvider {...{ children }} />
+    : <ServerLocationProvider {...{ children, initialLocation }} />
+}
+
+function BrowserLocationProvider({ children }) {
+  const history = React.useMemo(createHistory, [])
+  const [location, setLocation] = React.useState(() => history.location)
+
+  React.useEffect(
+    () => history.listen(({ location }) => setLocation(location)),
+    [history]
+  )
+
+  const value = React.useMemo(
+    () => ({ location, navigate: history.navigate }),
+    [location, history]
+  )
+
+  return <locationContext.Provider {...{ value, children }} />
+}
+
+function ServerLocationProvider({ initialLocation, children }) {
+  const value = React.useMemo(
+    () => ({
+      location: initialLocation,
+      navigate: () => { throw new Error('You can not navigate on the server') }
+    }),
+    [initialLocation]
+  )
+
+  return <locationContext.Provider {...{ value, children }} />
+}
+
+function RootBaseContextProvider({ children }) {
+  const { navigate } = React.useContext(locationContext)
+  const value = React.useMemo(
+    () => ({
+      basePath: '',
+      baseParams: {},
+      navigate: (to, x) => navigate(resolve('', to), x),
+    }),
+    [navigate],
+  )
+  return <baseContext.Provider {...{ value, children }} />
+}
+
+function NestedBaseContexProvider({ routePath, params, createChildren }) {
+  const { basePath, baseParams, navigate } = React.useContext(baseContext)
+  const { pathname } = useLocation()
+
+  const value = React.useMemo(
+    () => {
+      const newBasePath = pathname === '/' ? '' : pathname.replace(new RegExp(`${params['*']}$`), '')
+      return {
+        // basePath: `${basePath}${routePath}`, // we might want to keep this route path
+        basePath: newBasePath,
+        baseParams: { ...baseParams, ...params },
+        navigate: (to, x) => navigate(resolve(newBasePath, to), x)
+      }
+    },
+    [basePath, routePath, navigate, params, baseParams]
+  )
+  const children = callOrReturn(createChildren, value.baseParams)
+
+  return <baseContext.Provider {...{ value, children }} />
+}
+
+function Routing({ routes, initialLocation }) {
+  const context = React.useContext(locationContext)
+  if (!context && !inBrowser && !initialLocation)
+    throw new Error(`Your need to supply an initial location on server side rendering`)
```

Comment:
There's a typo: "Your" should be "You".
Additionally, I feel like it would be more clear if it would explain what I would need to do:

`You'll need to add the "initialLocation" prop on the "LocationProvider" for server-side rendering`

Or something like that